File: ConsoleApp/src/main/java/demos/SpringDemo.java
package demos;

import SpringDemo.concert.*;
import config.AppConfig;
import org.springframework.context.ApplicationContext;
import org.springframework.context.annotation.AnnotationConfigApplicationContext;

public class SpringDemo {

    public static void aopDemo() {
        ApplicationContext context = new AnnotationConfigApplicationContext(AppConfig.class);
        Performance performance = context.getBean(Performance.class);
        performance.doIt();

        // 由于实现了“引入新功能”
        // 此处转型为新功能接口
        Encoreable encoreable = (Encoreable) performance;
        encoreable.performEncore();
    }

    public static void executionDemo() {
        ApplicationContext context = new AnnotationConfigApplicationContext(AppConfig.class);
        IExecution myExecution = context.getBean("myExecution", IExecution.class);
        myExecution.doIt();

        IExecution sheExecution = context.getBean("sheExecution", IExecution.class);
        sheExecution.doIt();
    }

}

File: ConsoleApp/src/main/java/demos/SpringDemo.kt
package demos

import SpringDemo.concert.*
import config.AppConfig
import org.springframework.context.ApplicationContext
import org.springframework.context.annotation.AnnotationConfigApplicationContext

object SpringDemo {

    fun aopDemo() {
        val context = AnnotationConfigApplicationContext(AppConfig::class.java)
        val performance = context.getBean(Performance::class.java)
        performance.doIt()

        // 由于实现了“引入新功能”
        // 此处转型为新功能接口
        val encoreable = performance as Encoreable
        encoreable.performEncore()
    }

    fun executionDemo() {
        val context = AnnotationConfigApplicationContext(AppConfig::class.java)
        val myExecution = context.getBean("myExecution", IExecution::class.java)
        myExecution.doIt()

        val sheExecution = context.getBean("sheExecution", IExecution::class.java)
        sheExecution.doIt()
    }

}